### PR TITLE
feat: standalone rest server

### PR DESCRIPTION
**Pull request by Skillz4Killz**
**Please describe the changes this PR makes and why it should be merged:**
This is a serverless rest package that allows you to have an entire REST handler done separately.

- Easily host on any serverless infrastructure.
  - Easy to use and setup with Cloudflare Workers(FREE FOR 100K requests per day!)
-	Freedom from global rate limit errors
	- As your bot grows, you want to handle global rate limits better. Shards don't communicate fast enough to truly handle it properly so this allows 1 rest handler across the entire bot.
	- In fact, you can host multiple instances of your bot and all connect to the same rest server.
	- Even your dashboard can share the same rest handler using this to make sure all rate limits are handled from all points
- REST does not rest!
  - Separate rest means if your bot for whatever reason crashes, your requests that are queued will still keep going and will not be lost.
  - Seamless updates! When you want to update and reboot the bot, you could potentially lose tons of messages or responses that are in queue. Using this you could restart your bot without ever worrying about losing any responses.
- Scalability! Scalability! Scalability!

This does not follow all the same principles of Discordeno main API guidelines. In Discordeno, one of the goals was to limit as many options as possible to make ease of use/entry as simple as possible. This on the other hands is meant only for extremely large bot developers and it is at a point where all these options are not just good to have but an essential requirement.

```ts
startRESTServer({
  port: 8000,
  authorization: "HTP_SERVER_PASS_HERE",
  token: "BOT_TOKEN_HERE",
  maxRetryCount: 10
})

cache.eventHandlers.fetchFailed = function (data) {
  // Send to sentry, log, add to analytics do whatever u like
  console.log(data);
}
```

**Status**

- [ ] Code changes have been tested against the Discord API, or there are no code changes

**Semantic versioning classification:**

- [ ] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/rest/server.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+/** Handler function for every request. Converts to json, verified authorization & requirements and begins processing the request */
+async function handlePayload(
+  request: ServerRequest,
+  options: RestServerOptions,
+) {
+  const buffer = await Deno.readAll(request.body);
+
+  // INSTANTLY IGNORE ANY REQUESTS THAT DON'T HAVE THE SECRET AUTHORIZATION KEY
+  const authorization = request.headers.get("authorization");
+  if (authorization !== options.authorization) return;
```
Comment (vladfrangu) [inline src/rest/server.ts:27]: Read buffer after auth

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/rest/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Discordeno Rest
+
+This is a serverless rest package that allows you to have an entire REST handler done separately.
```
Comment (ayntee) [inline src/rest/README.md:3]: Deno enforces consistency (`package/module/etc` -> `module`).

```suggestion
A standalone and server-less REST module with functionality of REST, independently.
```
Better wording.

File: src/rest/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+- Easily host on any serverless infrastructure.
+  - Easy to use and setup with Cloudflare Workers(FREE FOR 100K requests per day!)
```
Comment (ayntee) [inline src/rest/README.md:6]:
```suggestion
  - Easy to use and setup with Cloudflare Workers (FREE for 100K requests per day!)
```

File: src/rest/README.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+-	Freedom from global rate limit errors
+	- As your bot grows, you want to handle global rate limits better. Shards don't communicate fast enough to truly handle it properly so this allows 1 rest handler across the entire bot.
+	- In fact, you can host multiple instances of your bot and all connect to the same rest server.
+- REST does not rest!
+  - Separate rest means if your bot for whatever reason crashes, your requests that are queued will still keep going and will not be lost.
+  - Seamless updates! When you want to update and reboot the bot, you could potentially lose tons of messages or responses that are in queue. Using this you could restart your bot without ever worrying about losing any responses.
+- Scalability! Scalability! Scalability!
+
+## Notes
```
Comment (ayntee) [inline src/rest/README.md:15]: I really think this note is not needed...

File: src/rest/deps.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export * from "https://deno.land/std@0.81.0/http/server.ts";
```
Comment (ayntee) [inline src/rest/deps.ts:1]:
```suggestion
export * from "https://deno.land/std@0.83.0/http/server.ts";
```

File: src/rest/queue.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        if (urlResetIn) continue;
+
+        // IF A BUCKET EXISTS, CHECK THE BUCKET'S RATE LIMITS
+        const bucketResetIn = queuedRequest.payload.bucketID
+          ? checkRateLimits(queuedRequest.payload.bucketID)
+          : false;
+        // THIS BUCKET IS STILL RATELIMITED, RE-ADD TO QUEUE
+        if (bucketResetIn) continue;
+
+        // EXECUTE THE REQUEST
```
Comment (ayntee) [inline src/rest/queue.ts:39]: Why every comment in caps? lol

File: src/rest/queue.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+          );
+
+          if (response.status < 200 && response.status >= 400) {
+            cache.eventHandlers.error(
+              "httpError",
+              queuedRequest.payload,
+              response,
+            );
+
+            const error = response.status === HttpResponseCode.BadRequest
```
Comment (ayntee) [inline src/rest/queue.ts:74]: Oh god

File: src/rest/request.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+  }
+
+  startQueue();
+}
+
+/** Creates the request body and headers that are necessary to send a request. Will handle different types of methods and everything necessary for discord. */
+export function createRequestBody(queuedRequest: QueuedRequest) {
+  const headers: { [key: string]: string } = {
+    Authorization: queuedRequest.options.token,
+    "User-Agent": "DiscordBot (https://github.com/discordeno/discordeno, v10)",
```
Comment (ayntee) [inline src/rest/request.ts:43]: `util/constants.ts` has a user-agent, can we use that?

File: src/rest/queue.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                queuedRequest.options.maxRetryCount
+            ) {
+              cache.eventHandlers.retriesMaxed(queuedRequest.payload);
+              queuedRequest.request.respond(
+                {
+                  status: 200,
+                  body: JSON.stringify(
+                    {
+                      error:
+                        "The reqest was rate limited and it maxed out the retries limit.",
```
Comment (ayntee) [inline src/rest/queue.ts:121]:
```suggestion
                        "The request was rate limited and it maxed out the retries limit.",
```

**Review by ayntee — APPROVED** [review]
LGTM! :heart: Is this tested?

image
One check failing :eyes:

File: src/rest/queue.ts (hunk 6 above)
Comment (Skillz4Killz) [reply src/rest/queue.ts:39]: Cause i was experimenting with different comment styles and turns out while i was coding i actually liked the caps style. It makes it much more understandable. Code is much easierto readforme

File: src/rest/queue.ts (hunk 7 above)
Comment (Skillz4Killz) [reply src/rest/queue.ts:74]: ?

File: src/rest/queue.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+export function startQueue() {
+  // IF ALREADY PROCESSING CANCEL
+  if (cache.processingQueue) return;
+  // MARK AS PROCESSING
+  cache.processingQueue = true;
+  processQueue();
+}
+
+/** Processes the queue by looping over each path separately until the queues are empty. */
+export async function processQueue() {
```
Comment (ayntee) [inline src/rest/queue.ts:15]: Redundant `async` function

File: src/rest/request.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+import { ServerRequest } from "./deps.ts";
+import { startQueue } from "./queue.ts";
+import {
+  QueuedRequest,
+  RestServerOptions,
+  RunMethodOptions,
+} from "./types/mod.ts";
+
+/** Processes a request and assigns it to a queue or creates a queue if none exists for it. */
+export async function processRequest(
```
Comment (ayntee) [inline src/rest/request.ts:11]: Redundant `async` keyword.

File: src/rest/request.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        bucketID,
+      });
+    }
+  }
+
+  return ratelimited ? bucketID : undefined;
+}
+
+/** This wll create a infinite loop running in 1 seconds using tail recursion to keep rate limits clean. When a rate limit resets, this will remove it so the queue can proceed. */
+async function processRateLimitedPaths() {
```
Comment (ayntee) [inline src/rest/request.ts:144]: Redundant `async` keyword

File: src/rest/cache.ts [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+  processingQueue: false,
+  globallyRateLimited: false,
+  ratelimitedPaths: new Map(),
+  eventHandlers: {
+    // BY DEFAULT WE WILL LOG ALL ERRORS TO CONSOLE. USER CAN CHOOSE TO OVERRIDE
+    error: function (_type, error) {
+      console.error(error);
+    },
+    // PLACEHOLDERS TO ALLOW USERS TO CUSTOMIZE
+    fetching: function () {},
```
Comment (ayntee) [inline src/rest/cache.ts:14]:
```suggestion
    fetching() {},
```

File: src/rest/cache.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    fetched: function () {},
+    fetchSuccess: function () {},
+    fetchFailed: function () {},
+    globallyRateLimited: function () {},
+    retriesMaxed: function () {},
```
Comment (ayntee) [inline src/rest/cache.ts:19]:
```suggestion
    fetched: function () {},
    fetchSuccess() {},
    fetchFailed() {},
    globallyRateLimited() {},
    retriesMaxed() {},
```

**Review by ayntee — APPROVED** [review]
LGTM :partying_face:

**Pull request merged**